plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ServletClasspathProvider.java

```diff
@@ -16,56 +16,118 @@
 
 package com.google.cloud.tools.eclipse.appengine.localserver;
 
-import com.google.cloud.tools.appengine.api.AppEngineException;
-import com.google.cloud.tools.appengine.cloudsdk.CloudSdk;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.Library;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFactory;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFactoryException;
+import com.google.cloud.tools.eclipse.appengine.libraries.model.LibraryFile;
+import com.google.cloud.tools.eclipse.appengine.libraries.repository.ILibraryRepositoryService;
+import com.google.cloud.tools.eclipse.appengine.libraries.repository.LibraryRepositoryServiceException;
 import com.google.cloud.tools.eclipse.util.MavenUtils;
 import com.google.common.annotations.VisibleForTesting;
-
+import com.google.common.base.Preconditions;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.inject.Inject;
 import org.eclipse.core.resources.IProject;
-import org.eclipse.core.runtime.Path;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.RegistryFactory;
+import org.eclipse.e4.core.contexts.ContextInjectionFactory;
+import org.eclipse.e4.core.contexts.EclipseContextFactory;
+import org.eclipse.e4.core.contexts.IEclipseContext;
 import org.eclipse.jdt.core.IClasspathEntry;
-import org.eclipse.jdt.core.JavaCore;
 import org.eclipse.jst.server.core.RuntimeClasspathProviderDelegate;
 import org.eclipse.wst.server.core.IRuntime;
+import org.osgi.framework.FrameworkUtil;
 
 /**
  * Supply Java standard classes, specifically servlet-api.jar and jsp-api.jar,
  * to non-Maven projects.
+ * <p>
+ * The jars are resolved using {@link ILibraryRepositoryService}.
  */
 public class ServletClasspathProvider extends RuntimeClasspathProviderDelegate {
-  private CloudSdk cloudSdkForTesting;
+
+  private static final Logger logger = Logger.getLogger(ServletClasspathProvider.class.getName());
+
+  private Map<String, Library> libraries;
+
+  @Inject
+  private ILibraryRepositoryService service;
+
+  private IEclipseContext eclipseContextForTesting;
```

Comment:
Ideally we should avoid `@Inject`s with `IEclipseContext`s: the ideal that we strive for is to have classes with `@Inject` that can be satisfied by Guice, Spring, or E4 DI.  Here's what I'd do to clean this up:

  1. create a 0-arg constructor that obtains the service context and then injects this instance
  2. create a multi-arg constructor for testing purposes that passes in the testable objects

An even better solution for (1) is to move the bundle service-context stuff into an`IExecutableExtensionFactory` that's specified in the extension point.  We'd put this in the `.util` bundle.  Then our `plugin.xml` would look like:
```
<runtimeClasspathProvider
  class="com.google.cloud.tools.eclipse.util.registry.ServiceContextFactory:com.google.cloud.tools.eclipse.appengine.localserver.ServletClasspathProvider"
…
```
This `ServiceContextFactory` would lookup the service context for the bundle that defined the `IConfigurationElement`.